lizmap/modules/lizmap/lib/Request/WFSRequest.php

```diff
@@ -849,7 +849,36 @@ protected function getfeaturePostgres()
         return new OGCResponse(200, 'application/vnd.geo+json; charset=utf-8', (function () use ($q) {
             yield '{"type": "FeatureCollection", "features": [';
             $virg = '';
+            $davProfile = RemoteStorageRequest::getProfile('webdav');
+            $baseUri = '';
+            if ($davProfile) {
+                $baseUri = $davProfile['baseUri'];
+            }
             foreach ($q as $d) {
+                $geoJson = json_decode($d->geojson, true);
+                // replace webdavUrl with a generic url
+                if ($geoJson['properties']) {
+                    // get webDav fields
+                    $webDavConfiguration = $this->qgisLayer->getWebDavFieldConfiguration();
+                    foreach ($geoJson['properties'] as $key => $value) {
+                        if (array_key_exists($key, $webDavConfiguration) && $value) {
+                            // if the base path starts with base URI, replace it with a genric path
+                            if ($baseUri && strpos($value, $baseUri) === 0) {
+                                $geoJson['properties'][$key] = str_replace($baseUri, 'dav/', $value);
+                            } else {
+                                // set filename as value
+                                $pathInfo = pathinfo($value);
+                                if ($pathInfo['filename'] !== '') {
+                                    $geoJson['properties'][$key] = $pathInfo['filename'].'.'.$pathInfo['extension'];
+                                } else {
+                                    $geoJson['properties'][$key] = $pathInfo['basename'];
+                                }
+                            }
+                        }
+                    }
+                }
+                $d->geojson = json_encode($geoJson);
```

Comment:
This code is only used for PostgreSQL layers, since you wrote it in the `getfeaturePostgres()` method. Perhaps it could also be great to support other types of layers used in Lizmap (users can have a FGB or GeoPackage edited from QGIS Desktop and containing WebDav fields pointing to the same WebDav server configured in `profiles.ini.php`. Not a strong opinion about it, as it could be added in the future.

Be aware that if there are some issues with the built SQL query (PostgreSQL errors), or if the QGIS layer contains joined fields from another layer, the function `getfeaturePostgres()` fallbacks to `return $this->getfeatureQgis();`, so you are not 100% sure that the Webdav fields will be correctly parsed, even for PostgreSQL layers.

I think this code must be extracted in a new function `processFeatureGeoJSON` that we could use in both methods `getfeaturePostgres()` and return `getfeatureQgis()`

You should not do anything if there are no webdav fields. Please move 
`$webDavConfiguration = $this->qgisLayer->getWebDavFieldConfiguration();` before calling the new function `processGeoJSONFeature` and if possible do not check if there are webdav fields for each feature but before (I think you should have all information before entering the foreach). Something like : 

```php
// Not sure there must be a new method $this->hasSomeWebdavFields()
// if the test is simple
foreach($q as $d) {
    if ($this->hasSomeWebdavFields()) {
        // Use the function processFeatureGeoJSON containing your code (without getWebDavFieldConfiguration)
        $d->geojson = $this->processFeatureGeoJSON($d->geojson);
    }
    yield $virg.$d->geojson;
    $virg = ',';
}
```

If we would like to compute WebDav fields only for PostgreSQL, we could also transfer the logic in the `buildQueryBase` method ( https://github.com/3liz/lizmap-web-client/blob/ca669772529ca59249babff35f9c306e14b8d294/lizmap/modules/lizmap/lib/Request/WFSRequest.php#L505 ) which could use PostgreSQL functions to directly change the values. Not a strong opinion on this : this would be more performant (no need to json_decode/json_encode) but would prevent from using it for non PostgreSQL layers

Reply:
> Be aware that if there are some issues with the built SQL query (PostgreSQL errors), or if the QGIS layer contains joined fields from another layer, the function getfeaturePostgres() fallbacks to return $this->getfeatureQgis();, so you are not 100% sure that the Webdav fields will be correctly parsed, even for PostgreSQL layers.

Right. Thanks for pointing it out to me.

> I think this code must be extracted in a new function processFeatureGeoJSON that we could use in both methods getfeaturePostgres() and return getfeatureQgis()

I agree. For me the goal is to support WebDav fileds for all datasources available, non only PostgreSQL. 

Thanks

Reply:
@mdouchin,
`getfeatureQgis` returns directly `OGCResponse` object. In order to process the response body I would store it  in a variable, process it and then return new OGCResponse like in the `getfeaturePostgres` function. Follow this idea, I would create a new function that return OGCResponse that can be called by both getFeatureQgis and getFeaturePostgres, something like that:
```php
protected function sendOGCResponse($httpCode,$contentType,$data){
     //.....
     return new OGCResponse($httpCode,$contentType,$data);
}
```
Do you think this could be the right approach?

I'm wondering if I should call `$this->processFeatureGeoJSON($geoJson)` anyway and move inside it the webdav logics. What do you think? Might it be useful in the future to have it for purposes other than webdav?

Reply:
You should not store de response in a variable, because it can be very heavy. This is why we know stream the response from QGIS and send it back to the brower in chunk by also streaming the response.
If you want to modify the DAV url you need to use tools such as json-machine 
My colleague @rldhont will give more details

Reply:
As @mdouchin do not stored response in a variable. You have to work as a stream.

First of all, move this code in its own function `processFeatureProperties`

Then you have to use `json-machine` to update the stream provided in `getFeatureQgis`
```php

        // Else pass query to QGIS Server
        // Get remote data as stream
        $wfsResult = $this->request(true, true);
        if ($wfsResult->code >= 400) {
            return $rep;
        }
        if ($wfsResult->mime != 'application/vnd.geo+json; charset=utf-8') {
            return $rep;
        }
        if (!$this->hasSomeWebdavFields()) {
            return $wfsResult;
        }
        if ($wfsResult->code >= 400) {
            return $rep;
        }
        if ($wfsResult->mime != 'application/vnd.geo+json; charset=utf-8') {
            // Do not process other format ?
            return $rep;
        }

        $featureStream = \Psr7\StreamWrapper::getResource($wfsResult->getBodyAsStream());
        $features = \JsonMachine\Items::fromStream($featureStream, ['pointer' => "/features"]);
        return new OGCResponse(200, 'application/vnd.geo+json; charset=utf-8', (function () use ($q) {
            yield '{"type": "FeatureCollection", "features": [';
            $virg = '';
            foreach ($features as $feat) {
                if (property_exists($feat, 'properties')) {
                    $feat->properties = $this->processFeatureproperties($feat->properties);
                }
                yield $virg.json_encode($feat);
                $virg = ',';
            }
      );
```

Reply:
marked as resolved since this part is no longer needed.